File: core/src/main/java/com/lomekwi/cave/timeline/playback/Playhead.java
package com.lomekwi.cave.timeline.playback;

import com.google.common.eventbus.EventBus;

import org.jspecify.annotations.NonNull;

public class Playhead {

    private volatile long anchor;
    private volatile long frozenTime = 0L;
    private volatile boolean isPlaying = false;
    private final transient @NonNull EventBus projEventBus;

    public Playhead(@NonNull EventBus projEventBus) {
        this.projEventBus = projEventBus;
    }

    public void setPlaying(boolean playing) {
        if (playing == isPlaying) return;

        if (playing) {
            anchor = System.nanoTime() - frozenTime;
        } else {
            frozenTime = System.nanoTime() - anchor;
        }

        isPlaying = playing;
    }

    public boolean isPlaying() {
        return isPlaying;
    }

    public void seek(long time) {
        time *= 1000;

        if (isPlaying) {
            anchor = System.nanoTime() - time;
        } else {
            frozenTime = time;
        }

        projEventBus.post(SeekEvent.INSTANCE);
    }

    public long getTime() {
        return getNanoTime() / 1000;
    }

    private long getNanoTime() {
        if (isPlaying) {
            return System.nanoTime() - anchor;
        } else {
            return frozenTime;
        }
    }
}

File: core/src/main/scala/com/lomekwi/cave/timeline/playback/Playhead.scala
package com.lomekwi.cave.timeline.playback

import com.google.common.eventbus.EventBus

class Playhead(@transient private val projEventBus: EventBus) {

  @volatile private var anchor: Long = 0
  @volatile private var frozenTime: Long = 0L
  @volatile private var playing: Boolean = false

  def setPlaying(playing: Boolean): Unit = {
    if (playing == isPlaying()) return

    if (playing) {
      anchor = System.nanoTime() - frozenTime
    } else {
      frozenTime = System.nanoTime() - anchor
    }

    this.playing = playing
    projEventBus.post(PlayStateChangedEvent.INSTANCE)
  }

  def isPlaying(): Boolean = playing

  def seek(time: Long): Unit = {
    var t = time
    t *= 1000

    if (isPlaying()) {
      anchor = System.nanoTime() - t
    } else {
      frozenTime = t
    }

    projEventBus.post(SeekEvent.INSTANCE)
  }

  def getTime(): Long = getNanoTime() / 1000

  private def getNanoTime(): Long = {
    if (isPlaying()) {
      System.nanoTime() - anchor
    } else {
      frozenTime
    }
  }
}
